File: clojure/48.clj
;; The series, 1^1 + 2^2 + 3^3 + ... + 10^10 = 10405071317.
;;
;; Find the last ten digits of the series, 1^1 + 2^2 + 3^3 + ... + 1000^1000.

(load "util")

(defn euler-48 [n m]
  (let [xs (map #(exp % %) (range (inc n)))
        sum (reduce +' xs)]
    (reduce str (take-last m (digits-of sum)))))

(run (euler-48 1000 10))

File: haskell/48.hs
-- The series, 1^1 + 2^2 + 3^3 + ... + 10^10 = 10405071317.
--
-- Find the last ten digits of the series, 1^1 + 2^2 + 3^3 + ... + 1000^1000.

import Util

euler48 n m = lastDigits m . sum . map (\x -> x^x) $ [1..n]
  where lastDigits m = fromDigits . reverse . take m . reverse . toDigits

main = run $ euler48 1000 10
